### ISSUE TITLE
Admin dashboard for referral codes

### ISSUE THREAD

**Issue by joemasilotti** [joemasilotti/railsdevs.com#728]
> ~Note: This is currently blocked by #736.~

Now that referral codes are saved via #727, we need a way to generate unique ones for referrers.

A PR addressing this issue should:

- [ ] Create a new admin dashboard at `/admin/referrals`
- [ ] List all of the users who have a referral code in a table
- [ ] Show how many referrals they have made
- [ ] Link to their referrals via `/admin/referrals/[:user_id]`
- [ ] Allow admins to create + update a new referral code for a user (auto generate the code itself)

This will require a new relationship (or more) and _might_ need a counter cache to keep the dashboard fast. Also, make sure to check out the other admin views as there are a lot of existing View Components that can be reused.

**Comment by franklinjosmell** [comment]
Hi Joe! I would like to work on this. The code have to based on some special pattern? (Like include the user/company initials) Would existits a character limit for the code?

**Comment by joemasilotti** [comment]
Thanks for the offer @franklinjosmell! Unfortunately this isn't ready to be worked on yet. I'll post here if/when I need community help on it.

**Comment by hafezbusra** [comment]
can i try work on this?

**Comment by joemasilotti** [comment]
Hey @hafezbusra - are you participating in [the hackathon](https://www.meetup.com/portland-ruby-brigade/events/290481353/) today? If so, message me in Slack and I can get you set up!

Otherwise, this is reserved for folks attending. But if it isn't claimed by the end of the day you can pick it up.

**Comment by hafezbusra** [comment]
im not on slack, can help send me the invitation at hafezbusra@gmail.com

**Comment by joemasilotti** [comment]
https://pdxruby.slack.com

**Comment by PatrickMcSweeny** [comment]
#736 would need to be completed before this one can be started

**Comment by joemasilotti** [comment]
Ah, great catch @PatrickMcSweeny! Let's hold off on this issue until that one is merged in.

**Comment by hafezbusra** [comment]
hi @joemasilotti is this not yet ready to work on?

**Comment by joemasilotti** [comment]
Hey @hafezbusra, this is now ready to be worked on. Do you want to take a pass?

**Comment by hafezbusra** [comment]
yup i would like to give it a try

**Comment by joemasilotti** [comment]
Awesome - all yours!

**Comment by joemasilotti** [comment]
Hey @hafezbusra, are you still working on this issue? Let me know if you are still planning on submitting a PR or if you need any help.

**Comment by hafezbusra** [comment]
oh sorry i didnt notice your reply, ya im planniing to submit pr for this. will finish this soon

**Comment by joemasilotti** [comment]
Great, thank you! Let me know if you run into any issues or need help.

**Comment by joemasilotti** [comment]
Can you send a screenshot? It sounds like you haven't forked the repo. After forking, push a branch there and create a PR on this repo.

**Comment by hafezbusra** [comment]
https://github.com/joemasilotti/railsdevs.com/pull/799, i havent finish the last task yet for allow admin to create a referral code and update. the parameter needed is only referral code right?

**Comment by joemasilotti** [comment]
This is looking good so far!

Product-wise, we also have to link to a show page that displays all the businesses and developers an individual has referred.

Do you want to split that and/or creating a new referral code to a new issue/PR? Or do you want to handle it within this one?

Also, can you please run `bundle exec standardrb --fix` to fix the lint errors? I'm also noticing some quote style changes to the locale file that should be reverted.

Thanks!

**Comment by hafezbusra** [comment]
I think I will do it in the same PR.

**Comment by joemasilotti** [comment]
Awesome.

> the parameter needed is only referral code right?

I think the requirement is actually a bit more complex. An admin needs to find a user in the system and create their referral code. So this would require some sort of search and then a button to create their code.

`/admin/users` has a search that works pretty well already. Copying something that works like that might be a good start.

Or... have the form require an exact match on an email address. Then require the admin to find it on their own. I'm fine with either for this PR - we can always come back and improve it later.

**Comment by hafezbusra** [comment]
> Product-wise, we also have to link to a show page that displays all the businesses and developer
do you mean like this
![Screenshot 2023-03-09 at 2 14 16 AM](https://user-images.githubusercontent.com/30567237/223796006-95a7f4bd-c34b-4e18-b15d-0860afc34694.png)
s an individual has referred.

**Comment by hafezbusra** [comment]
> Awesome.
> 
> > the parameter needed is only referral code right?
> 
> I think the requirement is actually a bit more complex. An admin needs to find a user in the system and create their referral code. So this would require some sort of search and then a button to create their code.
> 
> `/admin/users` has a search that works pretty well already. Copying something that works like that might be a good start.
> 
> Or... have the form require an exact match on an email address. Then require the admin to find it on their own. I'm fine with either for this PR - we can always come back and improve it later.

Ya. I think the first solution is better. I'll take a look at it and let you know. Thanks!

**Comment by joemasilotti** [comment]
Hey again @hafezbusra, do you still have appetite to get this over the finish line? I'm looking forward to seeing what you build!

**Comment by joemasilotti** [comment]
Closed via #799.
